experiment/image-bumper/main.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	imageRegexp = regexp.MustCompile(`\b(gcr\.io)/([a-z][a-z0-9-]{5,29}/[a-zA-Z0-9][a-zA-Z0-9_.-]+):([a-zA-Z0-9_.-]+)\b`)
+	tagRegexp   = regexp.MustCompile(`(v?\d{8}-(?:v\d(?:[.-]\d+)*-g)?[0-9a-f]{6,10}|latest)(-.+)?`)
+	tagCache    = make(map[string]string)
+)
+
+const (
+	imageHostPart  = 1
+	imageImagePart = 2
+	imageTagPart   = 3
+	tagVersionPart = 1
+	tagExtraPart   = 2
+)
+
+type manifest map[string]struct {
+	TimeCreatedMs string   `json:"timeCreatedMs"`
+	Tags          []string `json:"tag"`
+}
+
+func findLatestTag(imageHost, imageName, currentTag string) (string, error) {
+	k := imageHost + "/" + imageName + ":" + currentTag
+	if result, ok := tagCache[k]; ok {
+		return result, nil
+	}
+
+	currentTagParts := tagRegexp.FindStringSubmatch(currentTag)
+	if currentTagParts == nil {
+		return "", fmt.Errorf("couldn't figure out the current tag in %q", currentTag)
+	}
+	if currentTagParts[tagVersionPart] == "latest" {
+		return currentTag, nil
+	}
+
+	resp, err := http.Get("https://" + imageHost + "/v2/" + imageName + "/tags/list")
+	if err != nil {
+		return "", fmt.Errorf("couldn't fetch tag list: %v", err)
+	}
+
+	result := struct {
+		Manifest manifest `json:"manifest"`
+	}{}
+
+	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
+		return "", fmt.Errorf("couldn't parse tag information from registry: %v", err)
+	}
+
+	latestTag, err := pickBestTag(currentTagParts, result.Manifest)
+	if err != nil {
+		return "", err
+	}
+
+	tagCache[k] = latestTag
+
+	return latestTag, nil
+}
+
+func pickBestTag(currentTagParts []string, manifest manifest) (string, error) {
+	// The approach is to find the most recently created image that has the same suffix as the
```

Comment:
so the suffix here is 20191823-asdacdc?

Reply:
If your image is `gcr.io/k8s-testimages/kubekins-e2e:v20191823-afdacdc`, then the suffix is an empty string.

If your image is `gcr.io/k8s-testimages/kubekins-e2e:v20191823-afdacdc-experimental`, then the suffix is `-experimental`.

Reply:
Ah yes I confused myself with "prefix".